src/WithdrawRequestNFT.sol

```diff
@@ -23,21 +27,43 @@ contract WithdrawRequestNFT is ERC721Upgradeable, UUPSUpgradeable, OwnableUpgrad
 
     uint32 public nextRequestId;
     uint32 public lastFinalizedRequestId;
-    uint96 public accumulatedDustEEthShares; // to be burned or used to cover the validator churn cost
+    uint96 public DEPRECATED_accumulatedDustEEthShares; // to be burned or used to cover the validator churn cost
+
+    /// Stores the cached share price at the time of finalization for each batch
+    /// For a checkpoint index i, this batch includes requests:
+    /// `finalizationCheckpoints[i-1].lastFinalizedRequestId < requestId <= finalizationCheckpoints[i].lastFinalizedRequestId`
+    FinalizationCheckpoint[] public finalizationCheckpoints;
```

Comment:
I think you can get rid of binary search and just use a map of lastFinalizedRequestId -> finalizedPrice 
and add lastFinalizedRequestId in WithdrawRequest

Reply:
- The problem with this approach is that the oracle would have to update the lastFinalizedRequestId for each withdrawal request which is too expensive of an operational cost
- When using the binary search the oracle only has to push 1 element onto the `finalizationCheckpoints` vs having to update the `WithdrawRequest` for each request being finalized